File: lua/palette.lua
Palette = {}
Palette.__index = Palette

function Palette:new(maxIterations, image, redLevel, blueLevel)
	local self = setmetatable({}, Palette)
	self._image = image
	self._redLevel = redLevel
	self._blueLevel = blueLevel
	self.insideImageColor = self._image.allocateColor(0, 0, 0)
	self.scaleImageColor = self._image.allocateColor(0, 0, 0)
	self.factor = 1 / math.log(maxIterations) * 255;
	local cache = {};
	local i = 1
	while (i <= maxIterations) do
		cache[i] = -1
		i = i + 1
	end
	self._cache = cache
	return self
end

function Palette:getInsideColor()
	return self.insideImageColor
end

function Palette:getScaleColor()
	return self.scaleImageColor
end

function Palette:getColor(offset)
	if (self._cache[offset+1] >= 0) then
		return self._cache[offset+1]
	end
	local green = (math.log(offset) * self.factor);
	local ret = self._image.allocateColor(self._redLevel, green, self._blueLevel)
	self._cache[offset+1] = ret
	return ret
end

File: js/palette.js
function Palette(maxIterations, image, redLevel, blueLevel) {
	this._image = image;
	this._redLevel = redLevel;
	this._blueLevel = blueLevel;
	this.insideImageColor = this._image.allocateColor(0, 0, 0);
	this.scaleImageColor = this._image.allocateColor(0, 0, 0);
	this.factor = 1 / Math.log(maxIterations) * 255;
	var cache = [];
	var i;
	for (i = 0; i < maxIterations; i++) {
		cache[i] = -1;
	}
	this._cache = cache;

	this.getInsideColor = function() {
		return this.insideImageColor;
	}

	this.getScaleColor = function() {
		return this.scaleImageColor;
    }

    this.getColor = function(offset) {
		if (this._cache[offset] >= 0) {
			return this._cache[offset];
		}
		var green = (Math.log(offset) * this.factor) | 0;
		var ret = this._image.allocateColor(this._redLevel, green, this._blueLevel);
		this._cache[offset] = ret;
		return ret;
	}
}
